### ISSUE TITLE
clear dependent should be able to handle let bodies

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#11692]
#### Description of the problem
This should work, but it does not 
```coq
Goal forall x : Prop, let y := x in y -> True.
Proof. intros x y H. clear dependent x.
  (* hypothesis to clear is used in the conclusion (maybe indirectly) *)
```

#### Coq Version
8.11 / 8.12 / master

**Comment by herbelin** [comment]
See #11802. However, this does not fix #11691 (error message) nor #11689 (efficiency), which I don't know well how to address in pure Ltac.

**Comment by JasonGross** [comment]
Note that #11691 can be "fixed" simply my making the error message a bit more verbose: "hypothesis to clear is used in the conclusion (maybe indirectly) or is a section variable contained in the body of a definition which is used in a hypothesis"

#11689 can be fixed (at the cost of a bit of compatibility that probably (hopefully) no-one is taking advantage of) simply by replacing `match goal` with `lazymatch goal`, and I believe this is worth doing.
